backend/Shared/HitMeApp.Shared.Infrastructure/Integration/Memory/InMemoryIntegrationEventBus.cs

```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using HitMeApp.Shared.Infrastructure.Integration.Subscriptions;
+
+namespace HitMeApp.Shared.Infrastructure.Integration.Memory
+{
+    internal class InMemoryIntegrationEventBus
+    {
+        private static readonly Lazy<InMemoryIntegrationEventBus> _instance = new Lazy<InMemoryIntegrationEventBus>(() => new InMemoryIntegrationEventBus());
+
+        private readonly List<IIntegrationEventSubscription> _subscriptions = new List<IIntegrationEventSubscription>();
+
+        public static InMemoryIntegrationEventBus Instance => _instance.Value;
+
+        private InMemoryIntegrationEventBus()
+        {
+            // Hidden constructor due to singleton design pattern
+        }
+
+        public Task Publish<TIntegrationEvent>(params TIntegrationEvent[] events) where TIntegrationEvent : IntegrationEvent
+            => Publish(events?.AsEnumerable());
+
+        public Task Publish<TIntegrationEvent>(IEnumerable<TIntegrationEvent> events) where TIntegrationEvent : IntegrationEvent
+            => Task.WhenAll(events.Select(@event => Publish(@event)));
+
+        private Task Publish<TIntegrationEvent>(TIntegrationEvent @event) where TIntegrationEvent : IntegrationEvent
+        {
+            var handlers = _subscriptions.Where(subscription => subscription as IntegrationEventSubscription<TIntegrationEvent> is { })
```

Comment:
Would `subscription is IntegrationEventSubscription<TIntegrationEvent>` work in this case?

Reply:
Alternatively, with use of `as` we could chain it like so
```csharp
var handlers = _subscriptions.Select(subscription => subscription as IntegrationEventSubscription<TIntegrationEvent>)
.Where(subscription => subscription != null)
.Select(subscription => subscription.HandlerResolver().Handle(@event));
```

As right now we're making cast twice (with use of `as` operator and then using `Cast` method).

Reply:
I dunno about the corner cases, but I did similar things in the past and they worked quite well.
As I backup plan I store `Type`  in the subscription so we can do it differently.  Let's see about that after we introduce a second module to integrate with =)

Reply:
Double casting was indeed unnecessary. I simplified it according to your solution.